samples/subsys/usb/hid-keyboard/src/main.c

```diff
@@ -0,0 +1,247 @@
+/*
+ * Copyright (c) 2024 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <sample_usbd.h>
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/input/input.h>
+
+#include <zephyr/usb/usbd.h>
+#include <zephyr/usb/class/usbd_hid.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(main, LOG_LEVEL_INF);
+
+static const uint8_t __aligned(4) hid_report_desc[] = HID_KEYBOARD_REPORT_DESC();
+
+enum kb_leds_idx {
+	KB_LED_NUMLOCK = 0,
+	KB_LED_CAPSLOCK,
+	KB_LED_SCROLLLOCK,
+	KB_LED_COUNT,
+};
+
+static const struct gpio_dt_spec kb_leds[KB_LED_COUNT] = {
+	GPIO_DT_SPEC_GET_OR(DT_ALIAS(led0), gpios, {0}),
+	GPIO_DT_SPEC_GET_OR(DT_ALIAS(led1), gpios, {0}),
+	GPIO_DT_SPEC_GET_OR(DT_ALIAS(led2), gpios, {0}),
+};
+
+enum kb_report_idx {
+	KB_MOD_KEY = 0,
+	KB_RESERVED,
+	KB_KEY_CODE1,
+	KB_KEY_CODE2,
+	KB_KEY_CODE3,
+	KB_KEY_CODE4,
+	KB_KEY_CODE5,
+	KB_KEY_CODE6,
+	KB_REPORT_COUNT,
+};
+
+struct kb_event {
+	uint16_t code;
+	int32_t value;
+};
+
+K_MSGQ_DEFINE(kb_msgq, sizeof(struct kb_event), 2, 1);
+
+static uint8_t report[KB_REPORT_COUNT];
+static uint32_t kb_duration;
+static bool kb_ready;
+
+static void input_cb(struct input_event *evt)
+{
+	struct kb_event kb_evt;
+
+	kb_evt.code = evt->code;
+	kb_evt.value = evt->value;
+	if (k_msgq_put(&kb_msgq, &kb_evt, K_NO_WAIT) != 0) {
+		LOG_ERR("Failed to put new input event");
+	}
+}
+
+INPUT_CALLBACK_DEFINE(NULL, input_cb);
+
+static void kb_iface_ready(const struct device *dev, const bool ready)
+{
+	LOG_INF("HID device %p interface is %s", dev, ready ? "ready" : "not ready");
+	kb_ready = ready;
+}
+
+static int kb_get_report(const struct device *dev,
+			 const uint8_t type, const uint8_t id, const uint16_t len,
+			 uint8_t *const buf)
+{
+	return 0;
+}
+
+static int kb_set_report(const struct device *dev,
+			 const uint8_t type, const uint8_t id, const uint16_t len,
+			 const uint8_t *const buf)
+{
+	for (unsigned int i = 0; i < ARRAY_SIZE(kb_leds); i++) {
+		if (kb_leds[i].port == NULL) {
+			continue;
+		}
+
+		(void)gpio_pin_set_dt(&kb_leds[i], buf[0] & BIT(i));
+	}
+
+	return 0;
+}
+
+/* Idle duration is stored but not used to calculate idle reports. */
+static void kb_set_idle(const struct device *dev,
+			const uint8_t id, const uint32_t duration)
+{
+	LOG_INF("Set Idle %u to %u", id, duration);
+	kb_duration = duration;
+}
+
+static uint32_t kb_get_idle(const struct device *dev, const uint8_t id)
+{
+	LOG_INF("Get Idle %u to %u", id, kb_duration);
+	return kb_duration;
+}
+
+static void kb_set_protocol(const struct device *dev, const uint8_t proto)
+{
+	LOG_INF("Protocol changed to %s",
+		proto == 0U ? "Boot Protocol" : "Report Protocol");
+}
+
+static void kb_output_report(const struct device *dev, const uint16_t len,
+			     const uint8_t *const buf)
+{
+	LOG_HEXDUMP_DBG(buf, len, "o.r.");
+	kb_set_report(dev, 2U, 0U, len, buf);
```

Comment:
I meant that most of the parameters passed to `kb_set_report` are actually unused and we may desist from passing them as function arguments because of that

Reply:
kb_set_report is the callback implementation for set_report, I changed it a bit, maybe it makes it clearer.

Reply:
I see you prefer to pass the parameters anyway. We can keep it as is.